### PR TITLE
Fixed Segment Class

### LINKED ISSUE #1098: Geom: Refactor `Segment` to fit New Geometry hierarchy
### Description of the task
Segment should be refactored to fit the new classes and hierarchy in ```new_geom```. Currently, the ```segment.h``` in ```new_geom``` is a direct copy of the one in the old geom directory and should be refactored. Possible refactoring is to have line segment inherit from spline.
 

### Acceptance criteria

- [ ] Refactor ```segment``` such that it fits in the new geometry hierarchy.
- [ ] Unit tests for any new logic / previously untested logic

### Blocked by:
#983 Segment may inherit Spline in the new geometry hierarchy

### PR DESCRIPTION
**Pull request by jonathanlew**
### Description
* Added util functions to Segment and Point
* Refactored Segment a bit
* Segment can no longer be degenerate

### Testing Done
moved some unit tests around

### Resolved Issues
resolves #1098
part of #1146

### Review Checklist

**_It is the reviewers responsibility to also make sure every item here has been covered_**

- [ ] **Function & Class comments**: All function definitions (usually in the `.h` file) should have a javadoc style comment at the start of them. For examples, see the functions defined in `thunderbots/software/geom`. Similarly, all classes should have an associated Javadoc comment explaining the purpose of the class.
- [ ] **Remove all commented out code**
- [ ] **Remove extra print statements**: for example, those just used for testing
- [ ] **Resolve all TODO's**: All `TODO` (or similar) statements should either be completed or associated with a github issue
- [ ] **Justify drops in code coverage**: If this PR results in a non-trivial drop in code coverage (a bot should post a coverage diagram as a comment), please justify why we can't test the code that's not covered.

### CODE REVIEW

**Comment by MathewMacDougall** [comment]
Nice, took a quick look and no major comments atm. _Thank you_ for fixing `getSegStart`.

Regarding #1146 I just want to clarify what we want to move from util->geom classes or not. IIRC the reason we left a lot of stuff in `util` originally is because it was easier to have all the same "types" of operations in the same place (eg. `intersects`), and I think because it made things a bit simpler without bloating each individual class? 

I do remember us making the decision at some point to only leave "trivial" functions as class members. This is the main reason i want to check if #1146 is going to invovle moving as much as we can or not? If we start mixing where `intersects` functions are we'd need to move everything into the classes to preferably be consistent.

**Comment by jonathanlew** [comment]
> Nice, took a quick look and no major comments atm. _Thank you_ for fixing `getSegStart`.
> 
> Regarding #1146 I just want to clarify what we want to move from util->geom classes or not. IIRC the reason we left a lot of stuff in `util` originally is because it was easier to have all the same "types" of operations in the same place (eg. `intersects`), and I think because it made things a bit simpler without bloating each individual class?
> 
> I do remember us making the decision at some point to only leave "trivial" functions as class members. This is the main reason i want to check if #1146 is going to invovle moving as much as we can or not? If we start mixing where `intersects` functions are we'd need to move everything into the classes to preferably be consistent.

There are some issues with circular dependencies that are part of the discussion. Worth taking this offline

File: src/software/geom/util.cpp
Comment (akhilveeraghanta) [inline src/software/geom/util.cpp:345]: über mega nit: just checked the spec, its technically not wrong to provide the largest segment first, because the spec only says "first segment" and "second segment"

File: src/software/new_geom/rectangle.cpp
Comment (akhilveeraghanta) [inline src/software/new_geom/rectangle.cpp]: just curious why are we replacing `distanceFromPoint` with a subtraction between two points  and a length check? 

arguably Id say `distanceFromPoint` is a bit more readable

File: src/software/new_geom/segment.cpp
Comment (akhilveeraghanta) [inline src/software/new_geom/segment.cpp]: should we call this `getIntersection` to differentiate it even further from just `intersects`? 

don't feel strongly about this, so this is fine as long as we think its clear what these functions do

File: src/software/new_geom/segment.cpp
Comment (akhilveeraghanta) [inline src/software/new_geom/segment.cpp]: What would happen if both segments were identical? The spec says we only return 2 points, worried in the case that they are equal, we might return 4 points?

File: src/software/new_geom/segment.h
Comment (akhilveeraghanta) [inline src/software/new_geom/segment.h:19]: missing `@param`

File: src/software/new_geom/segment.h
Comment (akhilveeraghanta) [inline src/software/new_geom/segment.h:40]: o -> f

**Review by akhilveeraghanta — COMMENTED** [review]
nice work overall, segment is very clean 🔥

**Comment by garethellis0** [comment]
> > Nice, took a quick look and no major comments atm. _Thank you_ for fixing `getSegStart`.
> > Regarding #1146 I just want to clarify what we want to move from util->geom classes or not. IIRC the reason we left a lot of stuff in `util` originally is because it was easier to have all the same "types" of operations in the same place (eg. `intersects`), and I think because it made things a bit simpler without bloating each individual class?
> > I do remember us making the decision at some point to only leave "trivial" functions as class members. This is the main reason i want to check if #1146 is going to invovle moving as much as we can or not? If we start mixing where `intersects` functions are we'd need to move everything into the classes to preferably be consistent.
> 
> There are some issues with circular dependencies that are part of the discussion. Worth taking this offline

I'm just going to hold off on reviewing until we resolve this one. Slack (#software)?

File: src/software/ai/navigator/path_planner/theta_star_path_planner.cpp
Comment (EvanMorcom) [inline src/software/ai/navigator/path_planner/theta_star_path_planner.cpp]: Since has_line_of_sight is immediately set to *true* at the start of each logical block here - can we initially set it to *true* and have a single if(p1 != p2) block?

**Review by EvanMorcom — COMMENTED** [review]
Lookin' :fire: 

Quick comment, will finish reviewing tomorrow

File: src/software/new_geom/rectangle.cpp
Comment (jonathanlew) [reply src/software/new_geom/rectangle.cpp]: I think we should only have one or the other. Which one do you prefer?

File: src/software/new_geom/segment.cpp
Comment (jonathanlew) [reply src/software/new_geom/segment.cpp]: intersection is consistent with the intersection file, so that's how I chose that

File: src/software/new_geom/segment.cpp
Comment (jonathanlew) [reply src/software/new_geom/segment.cpp]: I will add a test

File: src/software/new_geom/segment.h
Comment (jonathanlew) [reply src/software/new_geom/segment.h:40]: good catch

File: src/software/new_geom/segment.h
Comment (jonathanlew) [reply src/software/new_geom/segment.h:19]: added

File: src/software/ai/navigator/path_planner/theta_star_path_planner.cpp
Comment (jonathanlew) [reply src/software/ai/navigator/path_planner/theta_star_path_planner.cpp]: changed it

File: src/software/new_geom/segment.cpp
Comment (jonathanlew) [reply src/software/new_geom/segment.cpp]: fixed a bug good catch

File: src/software/geom/util.cpp
Comment (jonathanlew) [reply src/software/geom/util.cpp:345]: I'm confused by this...

Comment (akhilveeraghanta) [reply src/software/geom/util.cpp:345]: the comment says "Swap if initialized wrong", its not to wrong to initialize the other way

again this is an uber mega nit, so don't bother

File: src/software/new_geom/rectangle.cpp
Comment (akhilveeraghanta) [reply src/software/new_geom/rectangle.cpp]: offline discussion: we are going to do `p1.distance(p2)`

Comment (jonathanlew) [reply src/software/new_geom/rectangle.cpp]: fixed it

**Review by akhilveeraghanta — COMMENTED** [review]
LGTM, but this PR is probably blocked until we agree on the class vs util split

**Comment by jonathanlew** [comment]
> 
> 
> LGTM, but this PR is probably blocked until we agree on the class vs util split

Since our codebase is written in an object-oriented style, I prefer functions to be in classes as much as possible.

**Comment by garethellis0** [comment]
> > LGTM, but this PR is probably blocked until we agree on the class vs util split
> 
> Since our codebase is written in an object-oriented style, I prefer functions to be in classes as much as possible.

I still strongly disagree with this, I think it will be much simpler and more maintainable in the future to keep all functions (as much as reasonably possible) in separate files. OOP and functional programming are tools to be used when appropriate, not religions to adhere to at all costs. When someone is thinking about where a geometry function might be or where one should go, the last thing we want is for them to have to follow several rules to decide. While I definitely get that overall your proposal is technically more consistent, I think it is practically much more difficult to adhere to and use.

**Comment by jonathanlew** [comment]
> I still strongly disagree with this, I think it will be much simpler and more maintainable in the future to keep all functions (as much as reasonably possible) in separate files. OOP and functional programming are tools to be used when appropriate, not religions to adhere to at all costs. When someone is thinking about where a geometry function might be or where one should go, the last thing we want is for them to have to follow several rules to decide. While I definitely get that overall your proposal is technically more consistent, I think it is practically much more difficult to adhere to and use.

In either proposal, there will rules for where things should go, and the rules aren't immediately intuitive.
I'm not convinced that your proposal is easier to adhere to, easier to implement, and easier to use.

**Comment by MathewMacDougall** [comment]
> In either proposal, there will rules for where things should go, and the rules aren't immediately intuitive.
I'm not convinced that your proposal is easier to adhere to, easier to implement, and easier to use.

I think the proposal about keeping everything in separate util files is easier to adhere to because _it is consistent with expectations_. Cody and Kelvin both found it more "immediately obvious" where things would go in our last in-person meeting. For someone reading the code for the first time, I think it will be more obvious to them that all distance functions are in the `distance` file, all intersects functions are in `intersects` etc.

While your (@jonathanlew ) proposal is consistent with the proposed rule (only put functions that operate on `Point` and the geometry type itself in the class), I don't think the rule itself is consistent with people's expectations. As a result I think it will require more manual review to enforce because people are more likely to mess it up until they learn the rule (which they may not until they are explicitly told). On the other hand, I think everything in specific util files is more "self-enforcing" because it is more "obvious" on a first read.

File: src/software/new_geom/segment.cpp
Comment (akhilveeraghanta) [inline src/software/new_geom/segment.cpp:51]: should this be in `contains.cpp`?

**Review by akhilveeraghanta — CHANGES_REQUESTED** [review]
just one comment

File: src/software/new_geom/segment.cpp
Comment (akhilveeraghanta) [reply src/software/new_geom/segment.cpp:51]: talked to @jonathanlew offline, mb didn't realize there was another PR that is doing that

**Review by akhilveeraghanta — APPROVED** [review]
LGTM 🔥

**Comment by garethellis0** [comment]
I think I will end up re-implementing most of these changes after the `contains` stuff merges, the merge conflicts will be very nasty otherwise.

**Comment by jonathanlew** [comment]
@garethellis0 Probably worth re-doing than fixing merge conflicts

**Pull request closed without merging**
